cloudpathlib/azure/azblobclient.py

```diff
@@ -250,24 +250,32 @@ def _move_file(
 
         return dst
 
-    def _remove(self, cloud_path: AzureBlobPath, missing_ok: bool = True) -> None:
+    def _remove(self, cloud_path: AzureBlobPath, missing_ok: bool = True) -> None:  # type: ignore
+        container_client = self.service_client.get_container_client(cloud_path.container)
         file_or_dir = self._is_file_or_dir(cloud_path)
+
+        if not file_or_dir:
+            if missing_ok:
+                return
+
+            raise FileNotFoundError(f"File does not exist: {cloud_path}")
+
         if file_or_dir == "dir":
-            blobs = [
-                b.blob for b, is_dir in self._list_dir(cloud_path, recursive=True) if not is_dir
-            ]
-            container_client = self.service_client.get_container_client(cloud_path.container)
-            container_client.delete_blobs(*blobs)
-        elif file_or_dir == "file":
-            blob = self.service_client.get_blob_client(
-                container=cloud_path.container, blob=cloud_path.blob
+            blobs = [(blob, is_dir) for blob, is_dir in self._list_dir(cloud_path, recursive=True)]
+
+            # need to delete files first to allow deleting the folders
+            files = [blob.blob for blob, is_dir in blobs if not is_dir]
+            container_client.delete_blobs(*files)
+
+            # folders need to be deleted from the deepest to the shallowest
+            folders = sorted(
+                (blob.blob for blob, is_dir in blobs if is_dir and blob.exists()), reverse=True
```

Comment:
I'm a little worried about perf since `.exists` adds a network call. This means that for large file trees (broad or deep) we potentially add a lot of calls for "fake" folders that don't actually exist on storage but `_list_dir` returns to act like a file system. 

I believe that only explicitly created folders on blob storage (or maybe with certain parameters set) would still stick around without specific removal. To create these, you need to be using hierarchical namespaces/a Data Lake Gen2 Storage Account.

I think the fix may instead to be to use an azure-SDK API that will list all explicit blobs (files or folders) rather than `_list_dir` in determining what to remove. I'm not sure if `list_blobs` does that under accounts with the hierarchical namespaces, so I think we'd need to test that first to see if it works for your use cases.

Do you think you could dig in and see if those settings/accounts can repro the issue for you?

Reply:
> I'm a little worried about perf since .exists adds a network call. This means that for large file trees (broad or deep) we potentially add a lot of calls for "fake" folders that don't actually exist on storage but _list_dir returns to act like a file system.

I don't think the call to `exists` here is strictly necessary, but I might be misremembering.

> I believe that only explicitly created folders on blob storage (or maybe with certain parameters set) would still stick around without specific removal. To create these, you need to be using hierarchical namespaces/a Data Lake Gen2 Storage Account.

Hmm, on further investigation, it seems the "problematic" accounts are using hierarchical namespaces. The directories seems to be "sticky" even without being explicitly created, which is a bit surprising.

Reply:
That's very helpful to know, thanks. I'd love to support that feature of blob storage—I'll set one up on our test account and see how it works. We may need more fixes than this to make `cloudpathlib` play nice with hierarchical namespace enabled storage accounts.